raw/fetcher.go

```diff
@@ -105,148 +128,83 @@ func (f *Fetcher) FetchInvoices(ctx context.Context, shouldUpdateTimeToLatest bo
-				glog.V(2).Infof("Reported %d invoices (last timestamp %v)", cnt, invoice.Timestamp)
+			if cnt%ReportBatch == 0 {
+				glog.V(2).Infof("Reported %d %s (last timestamp %v)", cnt, entityName, entity.Timestamp)
 			}
 		}
 	}
 }
 
-// FetchForwards will fetch and report forwards
-func (f *Fetcher) FetchForwards(ctx context.Context, shouldUpdateTimeToLatest bool, from time.Time) {
-
-	ctx = metadata.AppendToOutgoingContext(ctx, "pubkey", f.PubKey, "clientType", fmt.Sprintf("%d", f.ClientType), "key", f.AuthToken)
-
-	defer glog.Warningf("FetchForwards stopped running")
-
-	if shouldUpdateTimeToLatest {
-		ts, err := f.AgentAPI.LatestForwardTimestamp(ctx, &agent.Empty{})
-
-		if err != nil {
-			glog.Warningf("Coud not get latest forward timestamps: %v", err)
-		}
-
-		if ts != nil {
-			t := time.Unix(0, ts.Timestamp)
-			glog.V(2).Infof("Latest forwards timestamp: %v", t)
-			if t.After(from) {
-				from = t
-			}
-		}
+// FetchInvoices will fetch and report invoices
+func (f *Fetcher) FetchInvoices(ctx context.Context, shouldUpdateTimeToLatest bool, from time.Time) {
+	getTime := func(ctx context.Context) (*agent.TimestampResponse, error) {
+		return f.AgentAPI.LatestInvoiceTimestamp(ctx, &agent.Empty{})
 	}
 
-	outchan := GetForwardsChannel(ctx, f.LightningAPI, from)
+	getChan := func(ctx context.Context, itf api.LightingAPICalls, from time.Time) <-chan api.RawMessage {
+		return GetInvoicesChannel(ctx, itf, from)
+	}
 
-	cnt := 0
+	pushData := func(ctx context.Context, data *agent.DataRequest) error {
+		_, err := f.AgentAPI.Invoices(ctx, data)
+		return err
+	}
 
-	b := backoff.NewExponentialBackOff()
-	b.MaxElapsedTime = 0
+	f.fetch(ctx, "FetchInvoices", "invoices", getTime, getChan, pushData, shouldUpdateTimeToLatest, from)
+}
 
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case forward := <-outchan:
-			data := &agent.DataRequest{
-				Timestamp: forward.Timestamp.UnixNano(),
-				Data:      string(forward.Message),
-			}
+// FetchForwards will fetch and report forwards
+func (f *Fetcher) FetchForwards(ctx context.Context, shouldUpdateTimeToLatest bool, from time.Time) {
 
-			err := backoff.RetryNotify(func() error {
-				_, err := f.AgentAPI.Forwards(ctx, data)
-				return makePermanent(err)
-			}, b, func(e error, d time.Duration) {
-				glog.Warningf("Could not send data to GRPC endpoint: %v %v", data, e)
-			})
+	getTime := func(ctx context.Context) (*agent.TimestampResponse, error) {
+		return f.AgentAPI.LatestForwardTimestamp(ctx, &agent.Empty{})
+	}
 
-			if err != nil {
-				glog.Warningf("Fatal error in FetchForwards: %v", err)
-				return
-			}
+	getChan := func(ctx context.Context, itf api.LightingAPICalls, from time.Time) <-chan api.RawMessage {
+		return GetForwardsChannel(ctx, itf, from)
+	}
 
-			cnt++
-			if cnt%10 == 0 {
-				glog.V(2).Infof("Reported %d forwards (last timestamp %v)", cnt, forward.Timestamp)
-			}
-		}
+	pushData := func(ctx context.Context, data *agent.DataRequest) error {
```

Comment:
I see no added value in wrapping this into new variables. getTime, getChan and pushData.